File: Runtime/python/Phrost/Id.py
import secrets
import struct
from typing import Tuple

class Id:
    """
    Provides static methods for generating and handling 16-byte (128-bit) IDs
    by splitting them into two 64-bit integers.
    """

    @staticmethod
    def generate() -> Tuple[int, int]:
        """
        Generates 16 random bytes and returns them as two 64-bit integers.

        Note: Assumes 64-bit integers are unsigned and little-endian,
        matching PHP's 'J' pack format on a typical x86/ARM machine.
        """
        # 1. Generates 16 random bytes
        b = secrets.token_bytes(16)

        # 2. Split into two 8-byte parts
        p1 = b[:8]  # First 8 bytes
        p2 = b[8:]  # Second 8 bytes

        # 3. Unpack each part as an unsigned 64-bit little-endian integer ('<Q')
        # This is the Python equivalent of PHP's unpack("J", ...)
        return (
            struct.unpack("<Q", p1)[0],
            struct.unpack("<Q", p2)[0]
        )

    @staticmethod
    def to_bytes(ints: Tuple[int, int]) -> bytes:
        """
        Converts the 2-part integer tuple back into the 16-byte binary string.
        """
        # Pack each integer as an unsigned 64-bit little-endian integer ('<Q')
        # This is the Python equivalent of PHP's pack("J", ...)
        p1 = struct.pack("<Q", ints[0])
        p2 = struct.pack("<Q", ints[1])

        return p1 + p2

    @staticmethod
    def to_hex(ints: Tuple[int, int]) ->

File: Runtime/lua/phrost/id.lua
local M = {}

-- Seed the random number generator
math.randomseed(os.time())

function M.generate()
    -- Lua 5.4 supports 64-bit integers natively.
    -- We generate random bytes and unpack them into integers.
    local bytes = ""
    for i = 1, 16 do
        bytes = bytes .. string.char(math.random(0, 255))
    end
    -- unpack as two 64-bit unsigned integers (<I8 is Little Endian unsigned 64-bit)
    local p1, p2 = string.unpack("<I8I8", bytes)
    return p1, p2
end

function M.to_hex(id1, id2)
    local bytes = string.pack("<I8I8", id1, id2)
    local hex = ""
    for i = 1, #bytes do
        hex = hex .. string.format("%02x", string.byte(bytes, i))
    end
    return hex
end

return M
